timemachine/potentials/potentials.py

```diff
@@ -49,46 +73,76 @@ def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Ar
 
 
 @dataclass
-class ChiralAtomRestraint(Potential):
+class ChiralAtomRestraint(ShapedIdxsPotential):
     idxs: NDArray[np.int32]
 
     def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Array:
         return chiral_restraints.chiral_atom_restraint(conf, params, box, self.idxs)
 
+    def idxs_shape(self):
+        return (-1, 4)
+
+    def params_shape(self):
+        return -1
+
 
 @dataclass
-class ChiralBondRestraint(Potential):
+class ChiralBondRestraint(ShapedIdxsPotential):
     idxs: NDArray[np.int32]
     signs: NDArray[np.int32]
 
     def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Array:
         return chiral_restraints.chiral_bond_restraint(conf, params, box, self.idxs, self.signs)
 
+    def idxs_shape(self):
+        return (-1, 4)
+
+    def params_shape(self):
+        return -1
+
 
 @dataclass
-class FlatBottomBond(Potential):
+class FlatBottomBond(ShapedIdxsPotential):
     idxs: NDArray[np.int32]
 
     def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Array:
         return bonded.flat_bottom_bond(conf, params, box, self.idxs)
 
+    def idxs_shape(self):
+        return (-1, 2)
+
+    def params_shape(self):
+        return (-1, 3)
+
 
 @dataclass
-class LogFlatBottomBond(Potential):
+class LogFlatBottomBond(ShapedIdxsPotential):
     idxs: NDArray[np.int32]
     beta: float
 
     def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Array:
         return bonded.log_flat_bottom_bond(conf, params, box, self.idxs, self.beta)
 
+    def idxs_shape(self):
+        return (-1, 2)
+
+    def params_shape(self):
+        return (-1, 3)
+
 
 @dataclass
-class PeriodicTorsion(Potential):
+class PeriodicTorsion(ShapedIdxsPotential):
```

Comment:
Should we add a Proper/Improper set of classes so we can identify these in a list of Potentials? Else you need the VacuumSystem to tell which is which (or munging parameters)?

Reply:
I thought about this but decided against it. I view the classes in this file to be "functional forms" (and shouldn't be carrying semantics specific to the system of interest).

Reply:
The specifics of the system seem valuable, whether or not it should be defined in this file. Mostly thinking down the line where we want to work with a list of potentials, where we don't have the `VacuumSystem` object (such as pickling InitialState).

Not a strong preference for this, just thinking of the future.